library/modules/bigip_gtm_dns_listener.py

```diff
@@ -0,0 +1,836 @@
+            return None
+        return self._values['auto_lasthop']
+
+
+class Changes(Parameters):
+    def to_return(self):
+        result = {}
+        try:
+            for returnable in self.returnables:
+                result[returnable] = getattr(self, returnable)
+            result = self._filter_params(result)
+        except Exception:
+            raise
+        return result
+
+
+class UsableChanges(Changes):
+    pass
+
+
+class ReportableChanges(Changes):
+    @property
+    def address(self):
+        return self._values['address']
+
+    @property
+    def mask(self):
+        return self._values['mask']
+
+    @property
+    def port(self):
+        return self._values['port']
+
+    @property
+    def vlans(self):
+        return self._values['vlans']
+
+    @property
+    def rules(self):
+        return self._values['rules']
+
+    @property
+    def vlans_enabled(self):
+        return flatten_boolean(self._values['vlans_enabled'])
+
+    @property
+    def vlans_disabled(self):
+        return flatten_boolean(self._values['vlans_disabled'])
+
+    @property
+    def advertise(self):
+        if self._values['advertise'] is None:
+            return None
+        return flatten_boolean(self._values['advertise'])
+
+    @property
+    def ip_protocol(self):
+        if self._values['ip_protocol'] is None:
+            return None
+        return self._values['ip_protocol']
+
+
+class Difference(object):
+    def __init__(self, want, have=None):
+        self.want = want
+        self.have = have
+
+    def compare(self, param):
+        try:
+            result = getattr(self, param)
+            return result
+        except AttributeError:
+            return self.__default(param)
+
+    def __default(self, param):
+        attr1 = getattr(self.want, param)
+        try:
+            attr2 = getattr(self.have, param)
+            if attr1 != attr2:
+                return attr1
+        except AttributeError:
+            return attr1
+
+    @property
+    def state(self):
+        if self.want.state == 'disabled' and self.have.enabled:
+            return dict(
+                disabled=True
+            )
+        elif self.want.state in ['present', 'enabled'] and self.have.disabled:
+            return dict(
+                enabled=True
+            )
+
+    @property
+    def vlans(self):
+        if self.want.vlans is None:
+            return None
+        elif self.want.vlans == [] and self.have.vlans is None:
+            return None
+        elif self.want.vlans == self.have.vlans:
+            return None
+        else:
+            return self.want.vlans
+
+    @property
+    def rules(self):
+        if self.want.rules is None:
+            return None
+        elif self.want.rules == [] and self.have.rules is None:
+            return None
+        elif self.want.rules == self.have.rules:
+            return None
+        else:
+            return self.want.rules
```

Comment:
change this look at other modules for examples, also rules does not return empty list but empty string so you have a bug here

Reply:
Thanks for pointing out the existent logic! 

I have implemented the same way as it is done for virtual server. 

This comment is addressed now.